taskvine/src/manager/vine_manager.c

```diff
@@ -1369,19 +1372,14 @@ static int enforce_waiting_fixed_locations(struct vine_manager *q)
 {
 	struct vine_task *t;
 	int terminated = 0;
-	int count;
-
-	count = task_state_count(q, NULL, VINE_TASK_READY);
-	while (count > 0) {
-		count--;
 
-		t = list_pop_head(q->ready_list);
+	PRIORITY_QUEUE_ITERATE(q->ready_tasks, t)
```

Comment:
This looks like a problem to me:
- PRIORITY_QUEUE_ITERATE is a linear search.
- inside that iteration priority_queue_remove does another linear search to find t and remove it.

Put those together and yout have a O(n^2) algorithm.

Can you design this so that the item (already found) can be removed without a linear search?

Reply:
This routine should also break out after q->attempt_schedule_depth iterations and set a cursor to return to. It causes just as much if not overhead as send_one_task when it iterates through the whole ready list.

Reply:
However the overhead of this function is only of concern if we are using fixed_location temp files. During our previous optimization effort we conceded that fixed_location will generally be inefficient.

Reply:
This logic is integrated into `priority_queue_scheduling_next`. Every time trying to check if a task is able to run, that function advances the `scheduling_cursor`, which is an attribute in the `struct priority_queue`. If it doesn't find any tasks in the current round, the scheduling cursor is automatically remembered and will iterate starting from the current state in the next round. If something interesting happens (`events > 0` in the main loop), `priority_queue_scheduling_reset` is called and the scheduling cursor is reset.

However I am not very sure if `events > 0` is enough for the purpose of restoring the scheduling cursor. Is this too strict or too lenient, or just enough? @colinthomas-z80

Reply:
I think `events > 0` is the right idea. From what I understand most of the things that increment the event counter are things that we are interested in, such as a task retrieval or new worker connecting. Is there anything that increments the event counter that would not have an impact on being able to schedule a new task?

Reply:
We have these events that cause the event counter to increment
- a new worker is connected
- a worker is removed
- a task is dispatched to a worker
- a task is removed from the ready queue for whatever reasons
- a task is retrieved
- a file is replicated
- `process_pending` returns 1

Reply:
I think that 4, 6 and 7 don't seem to have an impact on enabling a blocking task to be scheduled. I might be wrong because I'm not very familiar with when `process_pending` would return 1.